### PR TITLE
Supporting of scalate view

### PR DESCRIPTION
**Pull request by c0d3r85**
- Supporting of scalate view: http://scalate.github.io/scalate/documentation/user-guide.html#Views
- Synchronized scalate render context locale with action locale
- Updated eclipse plugin version

### CODE REVIEW

**Comment by ngocdaothanh** [comment]
Is #17 not relevant any more because of this PR?

**Comment by c0d3r85** [comment]
Yep

**Comment by ngocdaothanh** [comment]
> Supporting of scalate view

Can you explain a little about Scalate views? Is it different from Scalate templates?

**Comment by c0d3r85** [comment]
Basicaly, it's ordinary template, but default searching algorithm for template differs: model classname (+super class and i-faces) transforms to path (no scalate prefix).

**Comment by c0d3r85** [comment]
https://github.com/scalate/scalate/blob/91905610bad4279e3067e495279081154c1a228e/scalate-core/src/main/scala/org/fusesource/scalate/RenderContext.scala#L318
In future, i think we should refactor scalate engine for loading resources with impl of  org.fusesource.scalate.util.ResourceLoader

**Comment by ngocdaothanh** [comment]
Guide about view: http://scalate.github.io/scalate/documentation/user-guide.html#Views

File: src/main/scala/xitrum/view/Scalate.scala [hunk 1]
```diff
@@ -105,6 +105,57 @@ object ScalateEngine {
...
+  }
+
+  def viewTemplate(model: AnyRef, engine: STE, viewName: String, templateType: String) = {
+    if (Config.productionMode) {
+      getPrecompiledScalateViewInstance(model, viewName, templateType)
+    } else {
+      val uri = getScalateViewInstance(developmentTemplateDirUri, model, viewName, templateType)
+      engine.load(uri, Nil)
+    }
+
```
Comment (ngocdaothanh) [inline src/main/scala/xitrum/view/Scalate.scala]: Please remove this unnecessary empty line

File: src/main/scala/xitrum/view/Scalate.scala [hunk 2]
```diff
@@ -105,6 +105,57 @@ object ScalateEngine {
...
+  }
+
+  final val developmentTemplateDirUri = "src/main/scalate"
```
Comment (ngocdaothanh) [inline src/main/scala/xitrum/view/Scalate.scala]: Please move this up, together with `CLASS_RESOLVER`, `CONTEXT_BINDING_ID` etc.

And also use upper case for it, like `DEVELOPMENT_TEMPLATE_DIR_URI`, or maybe shorter: `DEV_TEMPLATE_DIR_URI`.

File: src/main/scala/xitrum/view/Scalate.scala [hunk 3]
```diff
@@ -178,35 +226,32 @@ class ScalateEngine(
...
-    currentAction.at.foreach { case (k, v) =>
-      if (k == ACTION_BINDING_ID || k == CONTEXT_BINDING_ID)
-        log.warn(
-          ACTION_BINDING_ID + " and " + CONTEXT_BINDING_ID +
-          " are reserved key names for action's \"at\""
-        )
-      else
-        attributes.update(k, v)
+    currentAction: Action, options: Map[String, Any]): ScalateRenderContext =
+    {
```
Comment (ngocdaothanh) [inline src/main/scala/xitrum/view/Scalate.scala]: You are using 2 levels of indent (4 spaces). Please change it back to 1 level of indent (2 spaces).

File: src/main/scala/xitrum/view/ScalateRenderContext.scala [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    val template = ScalateEngine.viewTemplate(model, engine, viewName, templateType)
+    using(model) {
+      withUri(template.getClass.getName) {
+        template.render(this)
+      }
+    }
+  }
+
```
Comment (ngocdaothanh) [inline src/main/scala/xitrum/view/ScalateRenderContext.scala]: Please remove this unnecessary empty line

**Comment by ngocdaothanh** [comment]
Note:

Xitrum has method `renderFragment` to render a template fragment (like in Ruby on Rails):
http://xitrum-framework.github.io/guide/3.28/html/en/action_view.html#render-fragment

**Comment by c0d3r85** [comment]
With ```renderFragment``` we have to manually assign ```at(entityName) = model``` or pass through options. Scalate view is more convinient for render list of models, imho.

P.S.: I've updated PR.

**Comment by ngocdaothanh** [comment]
From the perspective of users, how will users use this new Scalate view feature?
(What methods and with what parameters to call?)

**Comment by c0d3r85** [comment]
action view:
```scalate
<% val person = new Person(firstName = "foo", surname = "bar") %>
<% val vipPerson = new VipPerson(firstName = "foo", surname = "bar") %>
<% val models = List(person, vipPerson) %>
<p>Something...</p>

<% view(person) %>
<% view(vipPerson) %>

#for(model <- models)
<% view(model) %>
#end

<% view(person, "edit") %>

<p>... more stuff </p>
```

Person.index.ssp:
```scalate
<%@ val it: Person %>
name: ${it.firstName} ${it.surname}
```

VipPerson.index.ssp:
```scalate
<%@ val it: VipPerson %>
VIP name: ${it.firstName} ${it.surname}
```

Person.edit.ssp:
```scalate
<%@ val it: Person %>
edit name: ${it.firstName} ${it.surname}
```

**Comment by ngocdaothanh** [comment]
Thanks, the API looks nice.
I'll try it, if there's no problem I'll merge this PR.

I'm thinking about adding a new API `renderPartial` to Xitrum.
The name is from Ruby on Rails:
http://api.rubyonrails.org/classes/ActionView/PartialRenderer.html

**Comment by c0d3r85** [comment]
Sorry, but I didn't catch difference ```renderPartial```  with:
```scala
def renderFragment[T <: Action: Manifest](fragment: String, options: Map[String, Any])
```
Could you explain in details?

**Comment by ngocdaothanh** [comment]
`renderPartial` is just another name of method `view` in the example in your previous comment.

If we add `renderPartial` to Xitrum core as a public API method, when we add other template engines to Xitrum, the new template engines will also need to implement it.

Maybe I should not add it now.

**Comment by ngocdaothanh** [comment]
Note: For `models` in the above example, we can use `collection` instead of `view`:

```
collection(models)
```

**Comment by ngocdaothanh** [comment]
Merged, thanks.
I'll release a new version of xitrum-scalate and update Xitrum Guide.

**Comment by ngocdaothanh** [comment]
Xitrum Guide says:

```
In templates you can use all methods of the class `xitrum.Action <https://github.com/xitrum-framework/xitrum/blob/master/src/main/scala/xitrum/Action.scala>`_.
Also, you can use utility methods provided by Scalate like ``unescape``.
See the `Scalate doc <http://scalate.fusesource.org/documentation/index.html>`_.
```

So I won't need to update it.

I've just released xitrum-scalate 2.8.0 and xitrum 3.28.3:
https://groups.google.com/forum/#!topic/xitrum-framework/Px9mWezMuOs

**Pull request merged**